config-system.lisp

```diff
@@ -0,0 +1,157 @@
+(in-package #:config-system)
+
+(defparameter *config-vars* (make-hash-table)
+  "Map that holds all of the configuration variables for the system")
+
+(defstruct config-info
+  (name t :type symbol :read-only t)
+  (validator #'identity :type (function (*) boolean))
+  (default nil :read-only t)
+  (doc "" :type string :read-only t))
+
+(defun %full-symbol-string (symb)
+  "Get the full name of the symbol complete with the package name"
+  (declare (type symbol symb))
+  (let* ((pkg (symbol-package symb))
+         (pkg-name (package-name pkg))
+         (symb-name (symbol-name symb)))
+    (concatenate 'string pkg-name
+                 (multiple-value-bind (symb status) (find-symbol symb-name pkg-name)
+                   (declare (ignore symb))
+                   (if (eql status :internal)
+                       "::"
+                       ":"))
+                 symb-name)))
+
+(define-condition config-error (error) ())
+
+(define-condition invalid-datum-error (config-error)
+  ((place-symbol :initarg :place :initform nil
+		 :accessor invalid-datum-error-place)
+   (value :initarg :value :initform nil
+	  :accessor invalid-datum-error-value))
+  (:report
+   (lambda (c s)
+     (with-slots (place-symbol value) c
+       (format s "The value ~S is invalid for variable ~S." value place-symbol)))))
+
+(define-condition config-not-found-error (config-error)
+  ((place-symbol :initarg :place :initform nil
+                 :accessor config-not-found-error-place
+                 :type symbol)
+   (alternatives :initarg :alternatives :initform nil
+                 :accessor config-not-found-alternatives
+                 :type list))
+  (:report
+   (lambda (c s)
+     (with-slots (place-symbol alternatives) c
+       (if alternatives
+        (format s  "Setting ~A not found. Did you mean one of these? ~A"
+                (%full-symbol-string place-symbol) alternatives)
+        (format s  "Setting ~A not found." (%full-symbol-string place-symbol)))))))
+
+(defun %generate-config-var-code (declare-type name default validator documentation)
+  (check-type documentation string)
+  (with-gensyms (default-value is-valid func)
+    `(progn
+       (,declare-type ,name ,default ,documentation)
+       (let* ((,default-value ,name)
+              (,func ,validator)
+              (,is-valid (funcall ,func ,default-value)))
+         (if ,is-valid
+             (setf (gethash (quote ,name) *config-vars*)
+                   (make-config-info :name (quote ,name) :default ,default-value
+                                     :validator ,func :doc ,documentation))
+             (error 'invalid-datum-error :place (quote ,name) :value ,default-value))))))
+
+(defmacro define-config-parameter (name default validator documentation)
```

Comment:
Using keyword parameters here might be nice to allow for future expansion.

Reply:
I agree with using keywords - I think it would be useful for things like changing the default database(s) or providing a different test function for `define-config-enum`, for example.

Reply:
Yeah, in particular I won't have separate macros for parameter/var, instead use keyword argument `:reinitialize` to use either defvar or defparameter.

I would also move documentation to a keyword argument and use the ['unsplice'](https://github.com/ruricolist/serapeum/blob/master/REFERENCE.md#unsplice-form) idiom ej. `,@(when documentation (list documentation))`. Or do we want to enforce a docstring for every configuration option?

> `define-config-enum`

I think that the valid values are better left as a keyword argument not a new kind of config parameter. ej. `(defconfig theme :type '(member 'dark 'light))`

Reply:
I lean towards having documentation be optional, as by requiring it the parameters/vars that wouldnt have them would most likely just put an empty string, which feels wrong. This is purely personal opinion, not necessarily informed by logic.

Reply:
> I wouldn't have separate macros for parameter/var, instead use keyword argument :reinitialize to use either defvar or defparameter.

Good idea!

> Or do we want to enforce a docstring for every configuration option?

This was my intention, but it looks like I'm outvoted. Considering these are settings, there should be some type of blurb to go along with it. I can see the merits of not requiring it, and like the interface of using a keyword better anyways. I'll take it out.

> I think that the valid values are better left as a keyword argument not a new kind of config parameter

That does feel more natural. So instead of specifying a validation function, you would specify a [type specifier](http://www.lispworks.com/documentation/HyperSpec/Body/04_bc.htm) that would be used to make a new type? A validation method is required though, so I'm hesitant to put that in a keyword argument, although it is pretty easy to signal an error when one isn't included.

Reply:
In [defconfig](https://github.com/szos/defconfig) I've been using just a quoted list of items (or unquoted if its a function to generate such a list) which gets spliced into a call to member for the `:valid-values` key argument. I like the idea of using type specifiers, but id like to have the ability to just pass in a list of acceptable values alongside it. I see this being useful for things like `*input-window-gravity*` and the like which are just a list of keywords.

side note - I would love any and all feedback on defconfig, as its my first real system with a real purpose, and there are definitely design flaws.

Reply:
The[`member`](http://www.lispworks.com/documentation/HyperSpec/Body/t_member.htm)type specifier takes care of that, as long as the values can be compared with `eql`. You also have the `satisfies` specifier, which can take any predicate function.

Reply:
Ah! this thread has been quite educating for me! love it :smile:

Reply:
> This was my intention, but it looks like I'm outvoted.

I was just asking questions :grimacing: . I may have brought up the pattern out of reflex/custom. But upon thinking about it, forcing a docstring is a valid choice to make sure we don't slack with the documentation. I say we keep it as obligatory. 

> So instead of specifying a validation function, you would specify a type specifier that would be used to make a new type?

StumpWM already has its own deftype macro, define-stumpwm-type. One way to go about it would be to extend it to support validators. But that would be on a separate PR.